api/client/client.go

```diff
@@ -161,13 +171,158 @@ func (c *Client) RemoveAppToken(ctx context.Context, tokenID string, auth Auth)
 	return err
 }
 
-type tokenAuth struct {
-	secure bool
+// AddFileOptions defines options for adding a file.
+type AddFileOptions struct {
+	Progress chan int64
+}
+
+// AddFileOption specifies an option for adding a file.
+type AddFileOption func(*AddFileOptions)
+
+// WithProgress writes progress updates to the given channel.
+func WithProgress(ch chan int64) AddFileOption {
+	return func(args *AddFileOptions) {
+		args.Progress = ch
+	}
+}
+
+type addFileResult struct {
+	path path.Resolved
+	err  error
+}
+
+// AddFile uploads a file to the project store.
+func (c *Client) AddFile(
+	ctx context.Context,
+	projID string,
+	filePath string,
+	auth Auth,
+	opts ...AddFileOption,
+) (path.Resolved, error) {
+	args := &AddFileOptions{}
+	for _, opt := range opts {
+		opt(args)
+	}
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	stream, err := c.c.AddFile(authCtx(ctx, auth))
+	if err != nil {
+		return nil, err
+	}
+	if err = stream.Send(&pb.AddFileRequest{
+		Payload: &pb.AddFileRequest_Header_{
+			Header: &pb.AddFileRequest_Header{
+				Name:      filepath.Base(file.Name()),
+				ProjectID: projID,
+			},
+		},
+	}); err != nil {
+		return nil, err
+	}
+
+	waitCh := make(chan addFileResult)
+	go func() {
+		defer close(waitCh)
+		for {
+			rep, err := stream.Recv()
+			if err == io.EOF {
+				return
+			}
+			if err != nil {
+				waitCh <- addFileResult{err: err}
+				return
+			}
+			switch payload := rep.GetPayload().(type) {
+			case *pb.AddFileReply_Event_:
+				if payload.Event.Path != "" {
+					id, err := cid.Parse(payload.Event.Path)
+					if err != nil {
+						waitCh <- addFileResult{err: err}
+						return
+					}
+					waitCh <- addFileResult{path: path.IpfsPath(id)}
+				} else if args.Progress != nil {
+					select {
+					case args.Progress <- payload.Event.Bytes:
+					default:
+					}
+				}
+			case *pb.AddFileReply_Error:
+				waitCh <- addFileResult{err: fmt.Errorf(payload.Error)}
+				return
+			default:
+				waitCh <- addFileResult{err: fmt.Errorf("invalid reply")}
+				return
+			}
+		}
+	}()
+
+	buf := make([]byte, chunkSize)
+	for {
+		n, err := file.Read(buf)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		if err = stream.Send(&pb.AddFileRequest{
+			Payload: &pb.AddFileRequest_Chunk{
+				Chunk: buf[:n],
+			},
+		}); err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
```

Comment:
Should in both returns call `stream.CloseSend()`?

Reply:
Ah yeah, good catch.